### PR TITLE
Only CSI plugin can have a DataSource

### LINKED ISSUE #96225: Creating an intree PVC with datasource should fail
**What happened**:
In clusters that have intree drivers as the default storageclass, if you try to create a PVC with snapshot data source and forget to put the csi storageclass in it, then an empty PVC will be provisioned using the default storageclass.

**What you expected to happen**:
PVC creation should not proceed and instead have an event with an incompatible error, similar to how we check proper csi driver in the csi provisioner.

**How to reproduce it (as minimally and precisely as possible)**:
https://stackoverflow.com/questions/64681941/kubernetes-pvc-restored-from-volumesnapshot-is-empty

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by xing-yang**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #96225

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
Creating a PVC with DataSource should fail for non-CSI plugins.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
```docs

```

### CODE REVIEW

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 1]
```diff
@@ -1473,6 +1473,15 @@ func (ctrl *PersistentVolumeController) provisionClaimOperation(
 	// called from provisionClaim(), in this case, plugin MUST NOT be nil
 	// NOTE: checks on plugin/storageClass has been saved
 	pluginName := plugin.GetPluginName()
+	if pluginName != "kubernetes.io/csi" && claim.Spec.DataSource != nil {
```
Comment (jsafrane) [inline pkg/controller/volume/persistentvolume/pv_controller.go:1476]: Can you please add an unit test somewhere in provision_test.go? Check e.g. how is `claimWithAnnotation()` used there and create `claimWithVolumeSource()` based on it. The rest should be trivial.

Comment (xing-yang) [reply pkg/controller/volume/persistentvolume/pv_controller.go:1476]: Done.

**Pull request merged**